api/v1_claim_rewards.go

```diff
@@ -225,65 +206,169 @@ func getValidatorAttestation(args GetValidatorAttestationParams) (*SenderAttesta
+	validatorsNeeded := minVotes - len(excludedOperators)
+	attestations := make([]SenderAttestation, 0, validatorsNeeded+offset)
+
+	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
 	defer cancel()
-	innerGroup, _ := errgroup.WithContext(ctx)
+
+	// Start AAO attestation in parallel if needed
+	var aaoAttestation *SenderAttestation
+	var aaoErr error
+	aaoDone := make(chan struct{})
 
 	if !hasAntiAbuseOracleAttestation {
-		innerGroup.Go(func() error {
+		go func() {
+			defer close(aaoDone)
 			aaoClaim := rewardClaim.RewardClaim
 			aaoClaim.AntiAbuseOracleEthAddress = ""
 			getAntiAbuseAttestationParams := GetAntiAbuseOracleAttestationParams{
 				Claim:                   aaoClaim,
 				Handle:                  rewardClaim.Handle,
 				AntiAbuseOracleEndpoint: antiAbuseOracle.Endpoint,
 			}
-			aaoAttestation, err := getAntiAbuseOracleAttestation(getAntiAbuseAttestationParams)
-			if err != nil {
-				return err
-			}
-			attestations[0] = *aaoAttestation
-			return nil
-		})
+			aaoAttestation, aaoErr = getAntiAbuseOracleAttestation(getAntiAbuseAttestationParams)
+		}()
+	} else {
+		close(aaoDone)
 	}
 
-	for i, validator := range validators {
-		innerGroup.Go(func() error {
-			getValidatorAttestationParams := GetValidatorAttestationParams{
-				Validator:      validator,
-				Claim:          rewardClaim.RewardClaim,
-				UserEthAddress: rewardClaim.RecipientEthAddress,
-				Signature:      signature,
+	// Get validator attestations in parallel rounds
+	successfulValidators := 0
+	currentIndex := 0
+
+	for successfulValidators < validatorsNeeded && currentIndex < len(shuffled) {
+		// Select up to validatorsNeeded validators for this round
+		var candidateNodes []config.Node
+		var candidateIndices []int
+		candidateOwners := make(map[string]bool) // Track owners in this round
+
+		for i := currentIndex; i < len(shuffled) && len(candidateNodes) < validatorsNeeded; i++ {
+			node := shuffled[i]
+			// Skip if we've already used this owner globally or this specific validator is marked bad
+			if usedOperators[node.OwnerWallet] || badValidators[node.Endpoint] {
+				continue
+			}
+			// Skip if we've already picked this owner in this round
+			if candidateOwners[node.OwnerWallet] {
+				continue
 			}
-			validatorAttestation, err := getValidatorAttestation(getValidatorAttestationParams)
+			candidateNodes = append(candidateNodes, node)
+			candidateIndices = append(candidateIndices, i)
+			candidateOwners[node.OwnerWallet] = true
+		}
 
-			if err != nil {
-				return err
+		if len(candidateNodes) == 0 {
+			break // No more valid candidates
+		}
+
+		// Try to get attestations from all candidates in parallel
+		type validatorResult struct {
+			index       int
+			attestation *SenderAttestation
+			err         error
+		}
+
+		results := make([]validatorResult, len(candidateNodes))
+		validatorGroup := &errgroup.Group{}
+
+		for i, node := range candidateNodes {
+			i, node := i, node // capture loop variables
```

Comment:
interesting... was this a bug before? i wasn't doing it before it looks like

Reply:
this is actually a thing
https://mrkaran.dev/tils/go-loop-var/

in practice, not sure if it was impacting us